packages/rpc-core/src/rpc-methods/getProgramAccounts.ts

```diff
@@ -0,0 +1,185 @@
+    RpcResponse,
+    Slot,
+    U64UnsafeBeyond2Pow53Minus1,
+} from './common';
+
+type AccountInfoBase = Readonly<{
+    /** indicates if the account contains a program (and is strictly read-only) */
+    executable: boolean;
+    /** number of lamports assigned to this account */
+    lamports: LamportsUnsafeBeyond2Pow53Minus1;
+    /** pubkey of the program this account has been assigned to */
+    owner: Base58EncodedAddress;
+    /** the epoch at which this account will next owe rent */
+    rentEpoch: U64UnsafeBeyond2Pow53Minus1;
+    /** the data size of the account */
+    space: U64UnsafeBeyond2Pow53Minus1;
+}>;
+
+type AccountInfoWithDefaultData = Readonly<{
+    data: Base58EncodedBytes;
+}>;
+
+type AccountInfoWithBase58EncodedData_DEPRECATED = Readonly<{
+    data: Base58EncodedDataResponse;
+}>;
+
+type AccountInfoWithBase64EncodedData = Readonly<{
+    data: Base64EncodedDataResponse;
+}>;
+
+type AccountInfoWithBase64EncodedZStdCompressedData = Readonly<{
+    data: Base64EncodedZStdCompressedDataResponse;
+}>;
+
+type AccountInfoWithJsonData = Readonly<{
+    data:
+        | Readonly<{
+              /** Name of the program that owns this account. */
+              program: string;
+              parsed: unknown;
+              space: U64UnsafeBeyond2Pow53Minus1;
+          }>
+        // If `jsonParsed` encoding is requested but a parser cannot be found for the given
+        // account the `data` field falls back to `base64`.
+        | Base64EncodedDataResponse;
+}>;
+
+type AccountWithPubkey<TAccount extends AccountInfoBase> = Readonly<{
+    account: TAccount;
+    pubkey: Base58EncodedAddress;
+}>;
+
+type GetProgramAccountsMemcmpFilter = Readonly<{
+    offset: U64UnsafeBeyond2Pow53Minus1;
+    bytes: string;
+    encoding: 'base58' | 'base64';
+}>;
+
+type GetProgramAccountsDatasizeFilter = Readonly<{
+    dataSize: U64UnsafeBeyond2Pow53Minus1;
+}>;
+
+type GetProgramAccountsApiCommonConfig = Readonly<{
+    /** @default "finalized" */
+    commitment?: Commitment;
+    /** The minimum slot that the request can be evaluated at */
+    minContextSlot?: Slot;
+    /** filter results (up to 4 filters allowed) @see https://docs.solana.com/api/http#filter-criteria */
+    filters?: (GetProgramAccountsMemcmpFilter | GetProgramAccountsDatasizeFilter)[];
+}>;
+
+type GetProgramAccountsApiSliceableCommonConfig = Readonly<{
+    /** Limit the returned account data */
+    dataSlice?: DataSlice;
+}>;
+export interface GetProgramAccountsApi {
+    /**
+     * Returns the account information for a list of Pubkeys.
+     */
+    getProgramAccounts(
+        program: Base58EncodedAddress,
+        config: GetProgramAccountsApiCommonConfig &
+            GetProgramAccountsApiSliceableCommonConfig &
+            Readonly<{
+                encoding: 'base64';
+                withContext: true;
+            }>
+    ): RpcResponse<AccountWithPubkey<AccountInfoBase & AccountInfoWithBase64EncodedData>[]>;
```

Comment:
What if…

https://www.typescriptlang.org/play?ssl=20&ssc=1&pln=20&pc=9#code/JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyA5hGACoAWoxAPOQMLrgQAeYTIMwxybkIACYBnHAHdgYSp0jsA-AC5kAI0wAbCHBABfAHwAKBM27FFyRs1kdjPZADIcBIkSMBbV5NcRwZ4WCg0ADROyNoAlEoWLOycJgDaAOQSUjL8CQC6fOzeIsj+AK4QIchyyCD5rsrQcenFSuWV0ATaBK34RiB+yHDIALzI+uVqat2i+SAA1iDoYiCjaJhhAHSkFNQgxPrYydKW-EoFEOEA3AQA9GfOyAB6cm0dXcp9A0MjcGOT07PzGOjLq1QaFsdql2Ep4GphEcwqd8Bcrrd7swughnoN8sN5uMpjM5u8Fn8VmRARstm4PGAvOAlAkAA5QdBIYRQwQJE7nS7ORH4AgPMDIQRo15Yz64n6LIlrIEwjkIuRAA

Does that work?

Reply:
I suspect it won't - I did something similar here: https://solanalabs.slack.com/archives/C050QNCV6BU/p1691418848861819

[This playground works correctly](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcBKYDGbgGdIB2uwAPACoBqAhgDYCuwAfHALzrBUAmEBNiJAbwBQcUXEw8YoGAC52XHn0Eixq3DQiy4BOgFsARsCgBuFaIC+jU6oButBnMr3gpy6aFIUcAOIAFAIIAwhC6ujzBBABmAJYA5uQA6tEwABYRUiAwzGw4Crz8wqoSocm6wAQwAPxyuDBQ0QSx1mIA7slpktLVcGRJqenSrlZCHsioGNh4hMQAkpF9HRXSie0DmQA0PTMEBEbZPQtr8NLlnLgIUAxwlehYOPg8xOTbu1DMji9G7pF0BJgw0R4PgCK36nUycBOBDOcH0EAgNA4BFYcEitGIjAAFGZVOIeDFYt0-EEQmECBECaDFhkskIAJRyCb3abAOaHcEwKlHTY6Ax7OCFMSpKAQFraYBigCiUBFUExdNcIwkRHgBAgFLicl5higKOJ8tMytq2ggvioUCoui1eh1eoCmIExV0pXKWgA5FQ3eYFUIjfA2mClplGXcpo9SNr+Wx9QIA9TpHI6gxvYbHqr1RzrXzddH7bHVpnUejgCmhEA)

But when I used the same pattern in this codebase it wouldn't work, and just treated the output as like `number | RpcResponse<number>` for any input

I think it might be the same underlying issue as https://github.com/solana-labs/solana-web3.js/issues/1280 but I'm not sure

Reply:
I fucked around and found out (that there's a TypeScript bug). https://github.com/solana-labs/solana-web3.js/pull/1531